pkg/flow/internal/controller/module_references.go

```diff
@@ -0,0 +1,77 @@
+package controller
+
+import (
+	"strings"
+
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/parser"
+)
+
+type ModuleReference struct {
+	componentName string
+	importLabel   string
+	declareLabel  string
+	importNode    *ImportConfigNode
+	declareNode   *DeclareNode
+}
+
+// This function will parse the provided river content and collect references to known modules.
+func GetModuleReferences(
+	content string,
+	importNodes map[string]*ImportConfigNode,
+	declareNodes map[string]*DeclareNode,
+	parentModuleDefinitions map[string]string,
+) ([]ModuleReference, error) {
+
+	uniqueReferences := make(map[string]ModuleReference)
+	err := getModuleReferences(content, importNodes, declareNodes, uniqueReferences, parentModuleDefinitions)
+	if err != nil {
+		return nil, err
+	}
+
+	references := make([]ModuleReference, 0, len(uniqueReferences))
+	for _, ref := range uniqueReferences {
+		references = append(references, ref)
+	}
+
+	return references, nil
+}
+
+func getModuleReferences(
+	content string,
+	importNodes map[string]*ImportConfigNode,
+	declareNodes map[string]*DeclareNode,
+	uniqueReferences map[string]ModuleReference,
+	parentModuleDefinitions map[string]string,
+) error {
+
+	node, err := parser.ParseFile("", []byte(content))
+	if err != nil {
+		return err
+	}
```

Comment:
I'm very uneasy with how much code is related to passing around strings and parsing them later. I would vastly prefer us to change things so that we passed around parsed AST blocks.

Reply:
Good idea, in this [commit](https://github.com/grafana/agent/pull/5968/commits/8f50ca1839f9e806aa99f30610711ab3a84cf070) I changed the code to use the Declare type instead of the plain content string. This way I can re-use the AST when searching for module references within a declare block. I also think that it makes the code clearer

Reply:
The declares are passed as pointers to avoid copying the string. I don't check for nil pointers because in the case when this would happen (which is not possible now), then it would be ok to panic because there would have been a programming error that would have allowed it and it should be fixed (no point to recover).